File: 0046-permutations/0046-permutations.java
class Solution {

    List<Integer> numbers;
    List<List<Integer>> ans;

    public void permutation(List<Integer> curr, int len, int[] nums) {
        if(curr.size() == len) {
            ans.add(new ArrayList<>(curr));
            return;
        }

        for(int i = 0; i < len; i++) {
            if(curr.contains(nums[i])) continue;
            curr.add(nums[i]);
            permutation(curr, len, nums);
            curr.remove(curr.size() - 1);
        }
    }

    public List<List<Integer>> permute(int[] nums) {
        numbers = new ArrayList<>();
        ans = new ArrayList<>();
        int len = nums.length;
        permutation(new ArrayList<>(), len, nums);
        return ans;
    }
}

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:

    void backtrack(vector<vector<int>>& ans, vector<int>& curr, vector<int>& nums, vector<bool>& cache) {
        if(curr.size() == nums.size()) {
            ans.push_back(curr);
            return;
        }

        for(int i = 0; i < nums.size(); i++) {
            if(cache[i]) continue;

            cache[i] = true;
            curr.push_back(nums[i]);
            backtrack(ans, curr, nums, cache);
            curr.pop_back();
            cache[i] = false;
        }
    
    }

    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> curr;
        vector<bool> cache(nums.size(), false);
        backtrack(ans, curr, nums, cache);
        return ans;
    }
};